### ISSUE TITLE
oidcstate Cookie not set

### ISSUE THREAD

**Issue by danutzplusplus** [thinktecture/Thinktecture.IdentityServer.v3.Samples#65]
Hello,

I've a question regarding the OpenID Http Module, and a specific part of its code. It has two parts which read and write a ProtectedCookie, containing the state and return url. The only issue is that the module seems to be crashing because it can't find the cookie.

And after investigating the source code, I notice that the cookie is instantiated, but not added to the response. Am I missing something? Shouldn't it be explicitly added to the response?

This is the httpmodule in question: https://github.com/thinktecture/Thinktecture.IdentityServer.v3.Samples/blob/master/source/Clients/Thinktecture.IdentityModel.Oidc/OpenIdConnectAuthenticationModule.cs

And it's instantiating the cookie, but not adding it:
var cookie = new ProtectedCookie(ProtectionMode.MachineKey);
cookie.Write("oidcstate", state + "_" + returnUrl, DateTime.UtcNow.AddHours(1));

context.Response.Redirect(authorizeUri.AbsoluteUri);

And when returning back, it crashes because it does not find it. 
var cookie = new ProtectedCookie(ProtectionMode.MachineKey);
var storedState = cookie.Read("oidcstate");
ProtectedCookie.Delete("oidcstate");

var separator = storedState.IndexOf('_');

And I checked the browser, and indeed the cookie is not there. Am I missing something? 

Thanks for the help
Dan

**Comment by brockallen** [comment]
Do you need to use the http module? In other words, why don't you use the OIDC Katana middleware? That's actually supported by Microsoft.

**Comment by danutzplusplus** [comment]
Well, I've got two scenarios, one client which is built over MVC 5, but I also have one over MVC4. And was wondering if there was an issue with the http module, or if I'm just missing something. And it's also a curiosity of mine as to how that cookie is set on the response, if it is, and I'm just not seeing it.

**Comment by brockallen** [comment]
Well, the http module we built was more of an educational exercise and we didn't really put as much effort into it as we do for other things that we intend to support.

**Comment by danutzplusplus** [comment]
Ok, cool, thanks. It's no problem really. I'm really appreciative for the work that you guys put in, and by no means was this a stab at the quality of your work. Just something that was puzzling me, that I maybe missed something.

**Comment by brockallen** [comment]
The other idea would be to look at the "MVC Code Flow Manual" sample from the samples repo -- in there we show how to do the whole OIDC protocol manually without protocol helper libraries.

**Comment by danutzplusplus** [comment]
Yeah, and that one works fine. Except for the little issue that the Katana middleware doesn't seem to validate the state parameter it generates to counter CSRF. At least I'm able to post to the callback action with any or no state. I've looked at the Microsoft.Owin.Security.OpenIdConnect dll in dotPeek, and there seems to be some code that seems to attempt to validate the state, but for some reason it's not running through that code. And I'd hate to have to figure out what they're doing to generate the CSRF token to validate it manually. That would be a pain.

**Comment by brockallen** [comment]
IIRC, the katana middleware does. You're saying it doesn't?

In any event, the "MVC code flow manual" doesn't use any katana middleware -- that's the point of it. To show how you'd do this without a protocol library.

**Comment by danutzplusplus** [comment]
Oh, I must have had an older version of the code flow manual sample.

But as far as the katana middleware, it's very weird. I've got the main controller action that I want to protect, it's decorated with Authorize, and I've configured the katana middleware to perform the whole openId code flow . And it works fine. 

On the redirection to the callback action (which has no Authorize restriction, right?) I take the auth code, and exchange it for an id token/access token handle. And I'm also getting the state and nonce posted, which is fine, even though, technically I shouldn't care about them since the katana middleware should have validated them before execution reaches the callback action, right?

Anyways, posting to the callback action, with or without the state or nonce in the request, doesn't seem to be validated by the katana middleware. I perform a manual barebones post to that callback action, and execution reaches the extraction of the auth code, which is null, and that value is attempted to be sent to the IdSrv for validation and exchange for a real token.

I was thinking that maybe it's something in the way I've configured the katana middleware. Here's how that looks:

app.UseOpenIdConnectAuthentication(new OpenIdConnectAuthenticationOptions
{
          ClientId = "clientid",
          Authority = "https://localhost:44333",
          RedirectUri = "https://localhost:44302/home/callback"
          ResponseType = "code",
          Scope = "openid profile",
          SignInAsAuthenticationType = "Cookies"
}

As I said, I've checked the OpenId dll from Microsoft in dotPeek, and I can see a check for the state value, but it's part of the last else in a series of if elses.

await this.get_Options().Notifications.MessageReceived(messageReceivedNotification);
if(((BaseNotification<OpenIdConnectAuthenticationOptions>) messageReceivedNotification).get_HandledResponse())
{
 ....
}
else if ((BaseNotification<OpenIdConnectAuthenticationOptions>) messageReceivedNotification).get_Skipped())
{
.....
}
else
{
//and here's where the state validation seems to be performed:
AuthenticationProperties properties = this.GetPropertiesFromState(openIdConnectMessage.get_State());
if(properties == null)
{
      LoggerExtensions.WriteWarning(this._logger, "The state field is missing or invalid.", new string[0]);
      authenticationTicket = (AuthenticationTicket)null;
      goto label_62;
}

Can you make more sense out of this?

**Comment by leastprivilege** [comment]
Try Microsoft's OWIN room on Jabbr.

**Comment by danutzplusplus** [comment]
Sorry to open this again, but I have some new info to share and some other question.
I've been talking with people in the Owin chat room on Jabbr, and after some further investigation it seems I should be handling the exchange of authcodes for tokens in the notification handlers that the katana OIDC client exposes (AuthorizationCodeReceived, SecurityTokenReceived, etc).

And I've tried that, but apparently that only seems to work if I explicitly request id_token among the other scopes). If I just request "code", as I did up until now, the notification handlers don't seem to be called (except fro MessageReceived, but that doesn't help).

I've then checked msdn for the AuthorizationCodeReceived handler, and apparently that only gets called after the security token validation (I imagine the identity token).

I've then checked the sample clients here (https://github.com/thinktecture/Thinktecture.IdentityServer.v3.Samples/tree/master/source/Clients). And noticed something interesting, which I didn't notice before. For the OWIN MVC clients there only seems to be samples for the hybrid (id_token and code) or the implicit flow. Is code flow not supported when requesting just an authcode which will be exchanged later on for the proper id token? I know the IdentityServer v3 does support it, but the katana OIDC middleware doesn't seem to trigger what I'd expect.
